LT_Types.ecl

```diff
@@ -0,0 +1,244 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2017 HPCC Systems®.  All rights reserved.
+############################################################################## */
+IMPORT $ AS LT;
+IMPORT ML_core;
+IMPORT ML_core.Types as CTypes;
+IMPORT LT.ndArray;
+t_Work_Item := CTypes.t_Work_Item;
+t_Count := CTypes.t_Count;
+t_RecordId := CTypes.t_RecordID;
+t_FieldNumber := CTypes.t_FieldNumber;
+t_FieldReal := CTypes.t_FieldReal;
+t_Discrete := CTypes.t_Discrete;
+t_TreeId := t_FieldNumber;
+Layout_Model := CTypes.Layout_Model;
+DiscreteField := CTypes.DiscreteField;
+NumericField := CTypes.NumericField;
+NumericArray := ndArray.NumericArray;
+
+/**
+  * Type definition module for Learning Trees.
+  */
+EXPORT LT_Types := MODULE
+
+  EXPORT t_NodeId := t_FieldNumber;
+
+  /**
+    * New Model Layout.  Should eventually be moved to ML_Core.
+    *
+    * Model is stored as an N-Dimensional numeric array via the
+    * ndArray.NumericArray class.
+    */
+  EXPORT Layout_Model2 := NumericArray;
+
+  /**
+    * Definition of the meaning of the indexes of the Model's
+    * NumericArray.  rfModInd1 enumerates the first index, which
+    * is used to determine which type of data is stored.
+    * - nodes stores the list of tree nodes that describes the forest.
+    *         The second index is just the sequential number of the node
+    *         The third index is enumerated below (rfModNodes3).
+    * - samples stores the set of sample indexes (i.e. ids) associated
+    *         with each treeId.
+    *         The second index represents the treeId.  The third index
+    *         represents the sample number. The value is the id of the
+    *         sample in the original training dataset.
+    *         [<samples, <treeId>, <sampleNum>] -> origId
+    * - classWeights (ClassificationForest only) stores the weights associated
+    *         with each class label.  The second index represents the class
+    *         label.  The value is the weight.  [<classWeights>, <classLabel>] -> weight
+    *
+    */
+  EXPORT rfModInd1 := ENUM(UNSIGNED4, nodes = 1, samples = 2, classWeights = 3);
```

Comment:
An UNSIGNED4 seems a waste for an enumeration type with 3 values.   Also, if this is likely to be in a file or a generated dataset, consider adding an Unknown=0 as zero is the default value for the field if someone used := [] on the right hand side.

Reply:
I'll add the Unknown = 0.  This is a value for an index in the ndArray.  I can't make it smaller since it needs to be the same size for all indexes and there can be large values for other members.

Reply:
Sounds like this is not an ENUM, but a set of constants.  I don't think that it will ever make much of a difference from a code generation perspective, but if we ever do more rigorous type checking (more semantics) it will cause problems.  Is there some useful behavior from using an ENUM construct?

Reply:
Good point.  I'll change those to constants.